### DISCUSSION TITLE
DI Examples not helpful or realistic

### DISCUSSION THREAD

**Discussion by MaxBrooks114** [quartznet/quartznet#1878] [General]
Trying to incorporate this package into our work and the only example for DI displays all the functionality from within the DI container. Not sure how this is helpful, considering the vast majority of jobs will be scheduled during the lifecycle of the app, from the service layer for instance, not before it even starts. There's no example on how to access the scheduler or scheduler factory from outside the container, as most people probably need.  Not only that, the scheduler itself is not even accessible/ instantiated until after the app runs - how is that helpful, when it can only be accessed from the DI builder as far as I can tell? I must be missing something, but the current documentation seems woefully lacking. I showed my lead and he was also confused and almost said don't use the package, which is a shame.

**Comment by lahma** [comment]
[The ASP.NET Core example ](https://github.com/quartznet/quartznet/blob/main/src/Quartz.Examples.AspNetCore/Pages/Index.cshtml.cs)site shows injection of `ISchedulerFactory` and with that you can get `IScheduler` which gives you all the power of the API to schedule/query/delete jobs and triggers.

**Comment by MaxBrooks114** [comment]
Thank you for the response. I have tried that and had no luck. If I use a named scheduler in my config, the factory returns null. If I try to get the default scheduler, it gets me a default scheduler with none of the config I set (using correct job store etc.)

```
builder.Services.Configure<QuartzOptions>(options =>
{
    options.Scheduling.IgnoreDuplicates = true;
    options.Scheduling.OverWriteExistingData = true;
});

builder.Services.AddQuartz(q =>
{
    q.SchedulerName = "JobRunnerScheduler"; 
    
    q.UseMicrosoftDependencyInjectionJobFactory();
    q.UseSimpleTypeLoader();
    q.UsePersistentStore(s =>
        {
            s.UseProperties = false;
            s.RetryInterval = TimeSpan.FromSeconds(15);
            s.UseSqlServer(sqlServer =>
            {
                sqlServer.ConnectionString = connectionString;
                // this is the default
                sqlServer.TablePrefix = "QRTZ_";
            });
            s.UseJsonSerializer();
            s.UseClustering(c =>
            {
                c.CheckinMisfireThreshold = TimeSpan.FromSeconds(20);
                c.CheckinInterval = TimeSpan.FromSeconds(10);
            });
        });
    q.UseDefaultThreadPool( tp =>
    {
        tp.MaxConcurrency = 10;
    });

});

builder.Services.AddQuartzServer(options =>
{
    // when shutting down we want jobs to complete gracefully
    options.WaitForJobsToComplete = true;
    options.AwaitApplicationStarted = false;

    
});

builder.Services.AddTransient<ISchedulerFactory, StdSchedulerFactory>();


builder.Services.AddScoped(x =>

       new JobRunner.Runner(x.GetService<ISchedulerFactory>(), x.GetService<IEmailSenderService>())
   );
```

```
public Runner(ISchedulerFactory schedulerFactory, IEmailSenderService emailSenderService)
       {
            _emailSenderService = emailSenderService;            

            _schedulerFactory = schedulerFactory;

            _scheduler = Task.Run(() => _schedulerFactory.GetScheduler("JobRunnerScheduler")).Result; //this returns null
       }
```

**Comment by lahma** [comment]
What is this:


```c#
builder.Services.AddTransient<ISchedulerFactory, StdSchedulerFactory>(); 
```

Quartz handles this for you, scheduler factory is anyway singleton, not transient. Try removing it.

**Comment by MaxBrooks114** [comment]
That is me being a dunce. It works. Thanks for the help. 

That being said injecting a "factory" into a class to "instantiate" a service sort of defeats the purpose of DI in the first place.

**Comment by lahma** [comment]
Well what scheduler instance should we inject then? There can be multiple. If it helps with the mental model think it as locator or accessor like there's `IHttpContextAccessor`.

**Comment by devbased** [comment]
@lahma
Can we use something like
`services.AddSingleton(provider => provider.GetRequiredService<ISchedulerFactory>().GetScheduler().GetAwaiter().GetResult());`
if app needs single scheduler?

**Comment by lahma** [comment]
@devbased probably yes, the default one isn't provided due to current integration being generic (also `GetAwaiter().GetResult() `could in very rare case actually do some async work).

**Comment by devbased** [comment]
I think one-time blocking async work is acceptable in general

**Comment by STeeL835** [comment]
To be honest, I think the documentation would benefit from a section that tells you how to use a scheduler if you use Quartz.NET in a ASP.NET app (which I would think is not a rare situation), or at least from a mention of examples (because I found this only after search, after I got an exception trying to run the app) 

I was reading through the web docs, and they do a good job trying to explain how all of it works, it created an impression that it is a full doc, but when it didn't mention this, I almost thought that the package is incomplete for using it with hosted services.

I'll see, maybe I'll have the time to make a PR to add this to docs, but my additions probably will be missing some details too

**Reply by STeeL835** [reply]
Also, it really isn't an accessor but a factory, because to actually get a scheduler you have to do some kind of async work, so you cannot get it in a ctor. Is it supposed to be created exclusively or be used exclusively? Can I get a seemingly created scheduler without await? It's a hosted service scheduler, I doubt it will be stopped without shutting down the app
